lightning/src/ln/onion_utils.rs

```diff
@@ -641,7 +641,41 @@ pub(super) fn process_onion_failure<T: secp256k1::Signing, L: Deref>(secp_ctx: &
 	} else { unreachable!(); }
 }
 
-/// Data decrypted from the onion payload.
+/// Used in the decoding of inbound payments' and onion messages' routing packets. This enum allows
+/// us to use `decode_next_hop` to return the payloads and next hop packet bytes of both payments
+/// and onion messages.
+enum Payload {
+	/// This payload was for an incoming payment.
+	Payment(msgs::OnionHopData),
+	/// This payload was for an incoming onion message.
+	Message(onion_message::Payload),
+}
+
+/// Similar to `Payload` above, this allows `decode_next_hop` to work for both payment and onion
+/// message forwards.
+enum NextPacketBytes {
+	/// The next hop's packet is for a payment, thus is fixed size.
+	Payment([u8; 20*65]),
+	/// The next hop's packet is for an onion message, thus is variable-length.
+	Message(Vec<u8>),
+}
+
+/// Data decrypted from an onion message's onion payload.
+pub(crate) enum MessageHop {
+	/// This onion payload was for us, not for forwarding to a next-hop.
+	Receive(onion_message::Payload),
+	/// This onion payload needs to be forwarded to a next-hop.
+	Forward {
+		/// Onion payload data used in forwarding the onion message.
+		next_hop_data: onion_message::Payload,
+		/// HMAC of the next hop's onion packet.
+		next_hop_hmac: [u8; 32],
+		/// Bytes of the onion packet we're forwarding.
+		new_packet_bytes: Vec<u8>,
+	},
+}
+
+/// Data decrypted from a payment's onion payload.
```

Comment:
Naming Bikeshedding Ep 27 Volume III : why not switch the name to `PaymentHop` to dissociate clearly from `MessageHop` ?